### PR TITLE
Highlighter-to-master

### PR DESCRIPTION
**Pull request by mnunberg**
This is the big PR including all highlighter related commits into master

### CODE REVIEW

File: src/byte_offsets.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
+#ifndef BYTE_OFFSETS_H
+#define BYTE_OFFSETS_H
+
+#include "redisearch.h"
+#include "varint.h"
+#include "rmalloc.h"
+
+typedef struct __attribute__((packed)) RSByteOffsetMap {
+  // ID this belongs to.
+  uint8_t fieldId;
```
Comment (dvirsky) [inline src/byte_offsets.h]: fieldId can be up to 1024, so this will overflow.

File: src/byte_offsets.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  uint32_t lastTokPos;
+} RSByteOffsetField;
+
+typedef struct RSByteOffsets {
+  // By-Byte offsets
+  RSOffsetVector offsets;
+  // List of field-id <-> position mapping
+  RSByteOffsetField *fields;
+  // How many fields
+  uint8_t numFields;
```
Comment (dvirsky) [inline src/byte_offsets.h]: we support up to 1024 fields, but only up to 32 indexed text fields. So as long as we apply this only to indexed text fields we are good.

File: src/redisearch.h [hunk 3]
```diff
@@ -135,7 +137,7 @@ typedef void (*RSFreeFunction)(void *);
 
 /* RSOffsetVector represents the encoded offsets of a term in a document. You can read the offsets
  * by iterating over it with RSOffsetVector_Iterate */
-typedef struct {
+typedef struct RSOffsetVector {
   char *data;
   size_t len;
```
Comment (dvirsky) [inline src/redisearch.h]: we can use uint32_t here and make sure we don't overflow in VVW_Write

File: src/result_termoffsets.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
+#ifndef RESULT_TERMOFFSETS_H
+#define RESULT_TERMOFFSETS_H
+
+#include "util/array.h"
+#include "redisearch.h"
+#include "byte_offsets.h"
+#include "fragmenter.h"
+
+typedef struct {
```
Comment (dvirsky) [inline src/result_termoffsets.h]: Document this. What is this struct? What is its purpose? etc

**Comment by mnunberg** [comment]
squashed

**Pull request closed without merging**